File: QueryandStore.js
const oracledb = require('oracledb');
const fs = require('fs');
const path = require('path');
const os = require('os');

console.log('Starting QueryandStore.js script.');

async function readFileContents(filePath) {
  try {
    const fileContents = fs.readFileSync(filePath, 'utf8');
    const lines = fileContents.split('\n');
    console.log(`File read successfully: ${filePath}`);
    return { connectionString: lines[0].trim(), query: lines.slice(1).join('\n').trim() };
  } catch (err) {
    console.error(`Failed to read file ${filePath}: ${err.message}`);
    throw err; 
  }
}

function parseConnectionString(connectionString) {
  try {
    const credentials = {
      user: '',
      password: '',
      remainingConnectionString: ''
    };
  
    const userIdMatch = connectionString.match(/User Id=([^;]*);/i);
    const passwordMatch = connectionString.match(/Password=([^;]*);/i);
  
    credentials.user = userIdMatch ? userIdMatch[1].trim() : '';
    credentials.password = passwordMatch ? passwordMatch[1].trim() : '';
    credentials.remainingConnectionString = connectionString
      .replace(/User Id=[^;]*;/i, '')
      .replace(/Password=[^;]*;/i, '')
      .replace(/Data Source=/i, '')
      .trim();
  
    console.log(`Connection string parsed successfully for user: ${credentials.user}`);
    return credentials;
  } catch (err) {
    console.error(`Failed to parse connection string: ${err.message}`);
    throw err;
  }
}

function toCSV(data) {
  try {
    const header = 'barcodes,carrierid,timestamp';
    const csvRows = data.map(row =>
      row.map(field =>
        `"${String(field).replace(/"/g, '""')}"`
      ).join(',')
    ).join(os.EOL);
    console.log(`Data converted to CSV format successfully`);
    return `${header}${os.EOL}${csvRows}`;
  } catch (err) {
    console.error(`Error converting data to CSV: ${err.message}`);
    throw err;
  }
}

function prepareStoragePath(basePath, subdir) {
  try {
    const now = new Date();
    const dateNow = now.toISOString().split('T')[0];
    const storagePath = path.join(basePath, subdir + dateNow);
    if (!fs.existsSync(storagePath)) {
      fs.mkdirSync(storagePath, { recursive: true }); 
      console.log(`Storage directory created: ${storagePath}`);
    }
    return storagePath;
  } catch (err) {
    console.error(`Error creating storage path ${subdir}: ${err.message}`);
    throw err;
  }
}

function writeResultToFile(csvData, filePath) {
  try {
    const now = new Date();
    const time = now.toISOString().substring(11, 19).replace(/:/g, '-');
    const baseFileName = path.basename(filePath, '.sql');
    const outputFileName = `${baseFileName}_${time}.csv`;
    const outputPath = path.join(filePath, outputFileName);
    fs.writeFileSync(outputPath, csvData);
    console.log(`Result written to CSV file ${outputPath}`);
  } catch (err) {
    console.error(`Failed to write result to file: ${err.message}`);
    throw err;
  }
}

async function mainThread(filePath) {
  
  let connection;
  try {
    const { connectionString, query } = await readFileContents(filePath);
    const { user, password, remainingConnectionString } = parseConnectionString(connectionString);
    connection = await oracledb.getConnection({ user, password, connectionString: remainingConnectionString });
    const result = await connection.execute(query);
    console.log(`Query executed successfully.`);
    const csvData = toCSV(result.rows);
    const storagePath = prepareStoragePath(__dirname, 'StorageResults/');
    writeResultToFile(csvData, storagePath);

  } catch (err) {
    console.error(`Error while processing file ${filePath}:`, err);
  } finally {
    if (connection) {
      try {
        await connection.close();
        console.log(`Connection closed for file.`);
      } catch (err) {
        console.error(`Error closing the connection for file ${filePath}:`, err);
      }
    }
  }
}

async function runQueriesInFolder(folderPath) {
  console.log(`Processing SQL files in folder: ${folderPath}`);
  const files = fs.readdirSync(folderPath);
  const queryPromises = files.filter(file => path.extname(file) === '.sql').map(file => mainThread(path.join(folderPath, file)));
  await Promise.all(queryPromises);
  const { exec } = require('child_process');
  console.log(`running S3 uploader...`);
  exec(`python3 s3_uploader.py`, (error, stdout, stderr) => {
    if (error) {
      console.error(`Error running S3 uploader script: ${error.message}`);
      return;
    }
    if (stderr) {
      console.error(`stderr from S3 uploader: ${stderr}`);
      return;
    }
    console.log(`stdout from S3 uploader: ${stdout}`);
  });
  console.log('Script execution complete.');
}

runQueriesInFolder(process.cwd());

File: QueryAndStore.py
import os
import sys
import csv
import cx_Oracle
import pathlib
from datetime import datetime
from subprocess import call

def read_file_contents(file_path):
    try:
        with open(file_path, 'r', encoding='utf-8') as file:
            lines = file.readlines()
        connection_string = lines[0].strip()
        headers = lines[1].strip().split(', ') 
        query = '\n'.join(lines[2:]).strip()  
        print(f"File read successfully: {file_path}")
        return connection_string, query, headers
    except Exception as e:
        print(f"Failed to read file {file_path}: {str(e)}")
        raise


def parse_connection_string(connection_string):
    try:
        user = ''
        password = ''
        dsn = ''
        if 'User Id=' in connection_string:
            user = connection_string.split('User Id=')[1].split(';')[0].strip()
        if 'Password=' in connection_string:
            password = connection_string.split('Password=')[1].split(';')[0].strip()
        if 'Data Source=' in connection_string:
            dsn = connection_string.split('Data Source=')[1].split(';')[0].strip()
        print(f"Connection string parsed successfully for user: {user}")
        return user, password, dsn
    except Exception as e:
        print(f"Failed to parse connection string: {str(e)}")
        raise

def to_csv(data, headers):
    try:
        with open('output.csv', 'w', newline='') as csvfile:
            csvwriter = csv.writer(csvfile, delimiter=',', quotechar='"', quoting=csv.QUOTE_MINIMAL)
            csvwriter.writerow(headers) 
            for row in data:
                csvwriter.writerow([str(field).replace('"', '""') for field in row])
        print("Data converted to CSV format successfully")
        return 'output.csv'
    except Exception as e:
        print(f"Error converting data to CSV: {str(e)}")
        raise


def prepare_storage_path(base_path, subdir):
    try:
        now = datetime.now()
        date_now = now.strftime('%Y-%m-%d')
        storage_path = pathlib.Path(base_path) / f"{subdir}{date_now}"
        storage_path.mkdir(parents=True, exist_ok=True)
        print(f"Storage directory created: {storage_path}")
        return storage_path
    except Exception as e:
        print(f"Error creating storage path {subdir}: {str(e)}")
        raise

def write_result_to_file(csv_data, file_path):
    try:
        now = datetime.now()
        time = now.strftime('%H-%M-%S')
        base_file_name = pathlib.Path(file_path).stem
        output_file_name = f"{base_file_name}_{time}.csv"
        output_path = pathlib.Path(file_path) / output_file_name
        with open(output_path, 'w') as f:
            f.write(csv_data)
        print(f"Result written to CSV file {output_path}")
    except Exception as e:
        print(f"Failed to write result to file: {str(e)}")
        raise

def main_thread(file_path):
    try:
        connection_string, query, headers = read_file_contents(file_path)
        user, password, dsn = parse_connection_string(connection_string)
        connection = cx_Oracle.connect(user, password, dsn)
        cursor = connection.cursor()
        result = cursor.execute(query)
        csv_data = to_csv(result.fetchall(), headers)
        storage_path = prepare_storage_path(os.getcwd(), 'StorageResults/')
        write_result_to_file(csv_data, storage_path)
    except Exception as e:
        print(f"Error while processing file {file_path}: {str(e)}")
    finally:
        if connection:
            connection.close()
            print(f"Connection closed for file {file_path}")

def run_queries_in_folder(folder_path):
    print(f"Processing SQL files in folder: {folder_path}")
    files = [f for f in os.listdir(folder_path) if f.endswith('.sql')]
    for file in files:
        main_thread(os.path.join(folder_path, file))
    print("running S3 uploader...")
    call(["python3", "s3_uploader.py"])


run_queries_in_folder(os.getcwd())
